pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java

```diff
@@ -56,16 +57,23 @@ public static BaseFilterOperator getLeafFilterOperator(PredicateEvaluator predic
     // Use inverted index if the predicate type is not RANGE or REGEXP_LIKE for efficiency
     DataSourceMetadata dataSourceMetadata = dataSource.getDataSourceMetadata();
     Predicate.Type predicateType = predicateEvaluator.getPredicateType();
-    if (dataSourceMetadata.hasInvertedIndex() && (predicateType != Predicate.Type.RANGE) && (predicateType
-        != Predicate.Type.REGEXP_LIKE)) {
-      if (dataSourceMetadata.isSorted()) {
+    if (dataSourceMetadata.hasInvertedIndex() && (predicateType != Predicate.Type.REGEXP_LIKE)) {
+      if (shouldUseSortedInvertedIndexOperator(dataSource, predicateType)) {
```

Comment:
I think checking `dataSourceMetadata.isSorted()` should be enough?

Reply:
The logic should be this right (also mentioned in the code comments)?
we can sorted inverted index if:
1. column is sorted AND
2. predicate is not RANGE OR predicate is RANGE but this physical plan is being built for offline segment

Reply:
If the column is sorted, we can use sorted index if there is no RANGE predicate -- we do this today.
If the column is sorted and there is RANGE predicate, use sorted index only for offline segments.

Reply:
Discussed offline.
dataSourceMetadata.isSorted() is enough since we dataSourceMetadata.hasInvertedIndex() will always be false for realtime segments so we don't need an enhanced check.

Reply:
You mean `dataSourceMetadata.isSorted()` is always false for realtime right?

Reply:
Exactly. Since MutableSegmentImpl calls the constructor of ColumnDataSource with isSorted as false.